### PR TITLE
auto activate or deactivate full frame depending on item type

### LINKED ISSUE #729: Item task: auto activate full frame mode
## Motivation

To enhance user experience, when displaying a task we should focus on the task − thus enter automatically “full frame” mode. 
“full frame” mode means the left menu is collapsed.

~However, when the user manually expands the left menu, we should disable that behavior for 2 hours, unless the user manually re-collapses the left menu.~

Fixing this issue will allow to delete [this branch](https://github.com/France-ioi/AlgoreaFrontend/compare/item-display-2?expand=1)

## Acceptance criteria (update 1/11, post discussion with Mathias)

- [x] 
  * When navigating to a task, **not using** the left nav for navigation
  * Then, the left menu collapses (if not already)
- [x] 
  * When navigating to a task, **using** the left nav for navigation
  * Then, the left menu stays visible
- [x] 
  * When navigating to a non-task (chapter, skill, group, ...) from a task 
  * Then, the left menu expends (if not already)
- [x] 
  * When navigating to a non-task (chapter, skill, group, ...) from a non-task 
  * Then, the left menu stays as it was
- [x] 
  * When using the collapse/expend button on the top of the left menu
  * Then, the left menu collapses/expends as expected

<details>
<summary>
Former acceptance criteria
</summary>

- [x] When:
- navigating to a chapter, a skill or a group

:arrow_right: full frame should not be activated
&nbsp;
- [x] When:
- navigating to a task

:arrow_right: full frame should be activated at task initialisation
&nbsp;
- [x] When:
- visiting a task
- expanding left menu manually
- visiting another task

:arrow_right: full frame should not be activated. This behavior should last 2 hours (add an escape hatch to test it on the branch)
&nbsp;
- [x] When:
- visiting a task
- expanding left menu manually
- collapsing left menu manually
- visiting another task

:arrow_right: full frame should remain activated
&nbsp;
- [ ] When:
- visiting a task
- navigate to a chapter
- expanding left menu manually
- visiting a task

:arrow_right: full frame should be activated at the (first and) last task initialisation
&nbsp;
</details>

### PR DESCRIPTION
**Pull request by ThomasJuster**
## Description

Fixes #729 
Closes #779 

## Notes

This PR _might_ evolve **slightly** when [this question](https://github.com/France-ioi/AlgoreaFrontend/issues/729#issuecomment-964154553) will be addressed.

Currently, I considered that courses ought to be treated just like tasks.

## Test cases

- [ ] Case 1:
  1. Given I am the usual user (to test access to a group page :wink:)
  2. When I navigate to [this chapter](https://dev.algorea.org/branch/feat/task-full-frame/en/#/activities/by-id/34689573454313988;path=4702,1352246428241737349,314613032161178344;attempId=0/details), [this skill](https://dev.algorea.org/branch/feat/task-full-frame/en/#/skills/by-id/3002;path=3000;parentAttempId=0/details) or [this group](https://dev.algorea.org/branch/feat/task-full-frame/en/#/groups/by-id/672913018859223173;path=52767158366271444/details)
  3. Then the left nav remains expanded
- [ ] Case 2:
  1. Given I am any user
  2. When I navigate to [this task](https://dev.algorea.org/branch/feat/task-full-frame/en/#/activities/by-id/784078458149384669;path=4702,1352246428241737349,314613032161178344,34689573454313988;attempId=0/details) or [this course](https://dev.algorea.org/branch/feat/task-full-frame/en/#/activities/by-id/837046206729127555;path=4702,1352246428241737349,314613032161178344;attempId=0/details)
  3. Then the left nav is collapsed
- [ ] Case 3:
  1. Given I am any user
  2. When I navigate to [this chapter](https://dev.algorea.org/branch/feat/task-full-frame/en/#/activities/by-id/34689573454313988;path=4702,1352246428241737349,314613032161178344;attempId=0/details)
  3. And I click on first item in right content (not the left nav)
  4. Then the left nav must be collapsed
- [ ] Case 4:
  1. Given I am any user
  2. When I navigate to [this task](https://dev.algorea.org/branch/feat/task-full-frame/en/#/activities/by-id/784078458149384669;path=4702,1352246428241737349,314613032161178344,34689573454313988;parentAttempId=0/details)
  3. And I click on next item in header
  4. Then the next item must be displayed and the left nav **remains** collapsed
  5. And I click on previous item in header
  6. Then the previous item must be displayed and the left nav **remains** collapsed
  7. And I click on parent (which is a chapter) in item header
  8. Then the parent chapter must be displayed and the left must be expanded
- [ ] Case 5:
  1. Given I am any user
  2. When I navigate to [this task](https://dev.algorea.org/branch/feat/task-full-frame/en/#/activities/by-id/784078458149384669;path=4702,1352246428241737349,314613032161178344,34689573454313988;parentAttempId=0/details) or [this chapter](https://dev.algorea.org/branch/feat/task-full-frame/en/#/activities/by-id/34689573454313988;path=4702,1352246428241737349,314613032161178344;parentAttempId=0/details)
  3. And I expand left nav (if not expanded)
  4. And I click on "Les boucles imbriquées" **in the left nav**
  5. Then the left nav must **remain** expanded
- [ ] Case 5:
  1. Given I am any user
  2. When I navigate to [this task](https://dev.algorea.org/branch/feat/task-full-frame/en/#/activities/by-id/784078458149384669;path=4702,1352246428241737349,314613032161178344,34689573454313988;parentAttempId=0/details) or [this chapter](https://dev.algorea.org/branch/feat/task-full-frame/en/#/activities/by-id/34689573454313988;path=4702,1352246428241737349,314613032161178344;parentAttempId=0/details)
  3. And I expand left nav
  4. Then the left must be toggled as expected: it collapses if it was expanded and vice-versa.
- [ ] Case 6:
  1. Given I am any user
  2. I visit [this chapter](https://dev.algorea.org/branch/feat/task-full-frame/en/#/activities/by-id/34689573454313988;path=4702,1352246428241737349,314613032161178344;attempId=0/details)
  3. I manually collapse the menu
  4. I visit the parent (or non-task sibling) through the header
  5. Then the left menu stays as it was: collapsed

## No regression tests

- [ ] Case 1:
  1. Given I am any user
  2. And I navigate to [this skill](https://dev.algorea.org/branch/feat/task-full-frame/en/#/skills/by-id/3002;path=3000;attempId=0/details)
  3. And I click on a sub skill or parent skill in right content (not left nav)
  4. Then the left nav must remain expanded
- [ ] Case 2:
  1. Given I am any user
  2. And I navigate to [this skill](https://dev.algorea.org/branch/feat/task-full-frame/en/#/skills/by-id/3002;path=3000;attempId=0/details)
  3. And I navigate using header with parent/prev/next buttons
  4. Then the left nav must remain expanded
- [ ] Case 3:
  1. Given I am the usual user
  2. And I navigate to [this group](https://dev.algorea.org/branch/feat/task-full-frame/en/#/groups/by-id/672913018859223173;path=52767158366271444/details)
  3. And I navigate using left nav or header buttons
  4. Then left nav must remain expanded

### CODE REVIEW

**Review by smadbe — CHANGES_REQUESTED** [review]
A case which does not respect
>  When navigating to a non-task (chapter, skill, group, ...) from a non-task
>  Then, the left menu stays as it was

- I visit a [chapter](https://dev.algorea.org/branch/feat/task-full-frame/en/#/activities/by-id/34689573454313988;path=4702,1352246428241737349,314613032161178344;attempId=0/details)
- I manually collapse the menu
- Then I visit the parent through the header.
- I would expect "the left menu stays as it was", but it auto-expends

**Review by smadbe — APPROVED** [review]

File: src/app/core/services/navigation/item-nav-tree.service.ts [hunk 1]
```diff
@@ -62,7 +62,7 @@ abstract class ItemNavTreeService extends NavTreeService<ItemInfo> {
     return {
       ...treeElement,
       title: details.title ?? '',
-      navigateTo: (): void => this.itemRouter.navigateTo(contentInfo.route),
+      navigateTo: (_path: string[], preventFullFrame = false): void => this.itemRouter.navigateTo(contentInfo.route, { preventFullFrame }),
```
[the remark this answers is not in the document]
Comment (ThomasJuster) [reply src/app/core/services/navigation/item-nav-tree.service.ts:65]: At first I thought the same :smile:, actually this is not exactly the left menu now, it's also used by the item-header

File: src/app/modules/item/pages/item-by-id/item-by-id.component.ts [hunk 2]
```diff
@@ -114,6 +118,19 @@ export class ItemByIdComponent implements OnDestroy {
...
       ).subscribe(() => this.reloadContent()),
+
+      this.itemDataSource.state$.pipe(
+        readyData(),
+        distinctUntilChanged((a, b) => a.item.id === b.item.id),
+        startWith(undefined),
+        pairwise(),
+        filter(([ previous, current ]) => !previous || isTask(previous.item) || !current || isTask(current?.item)),
+        map(([ , current ]) => current?.item),
+        filter(isNotUndefined),
```
[the remark this answers is not in the document]
Comment (ThomasJuster) [reply src/app/modules/item/pages/item-by-id/item-by-id.component.ts]: fixed

File: src/app/modules/item/pages/item-by-id/item-by-id.component.ts [hunk 3]
```diff
@@ -114,6 +118,19 @@ export class ItemByIdComponent implements OnDestroy {
         filter(mode => [ Mode.Normal, Mode.Watching ].includes(mode)),
         distinctUntilChanged(),
       ).subscribe(() => this.reloadContent()),
+
+      this.itemDataSource.state$.pipe(
+        readyData(),
+        distinctUntilChanged((a, b) => a.item.id === b.item.id),
+        startWith(undefined),
+        pairwise(),
+        filter(([ previous, current ]) => !previous || isTask(previous.item) || !current || isTask(current?.item)),
```
[the remark this answers is not in the document]
Comment (ThomasJuster) [reply src/app/modules/item/pages/item-by-id/item-by-id.component.ts]: fixed

**Pull request merged**